what happens on restart page (#56254)
kubernetes/website: content/en/docs/reference/node/what-happens-on-restart.md

Reviewer: Concepts we could mention in this page include:

- [graceful node shutdown](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
- Cordons (no concept page yet, so no hyperlink)
- Drains (no concept page yet, so no hyperlink)
- [Node heartbeats](https://kubernetes.io/docs/concepts/architecture/nodes/#node-heartbeats), and when they resume after a kubelet restart
- systemd heartbeats, and when these resume after a kubelet restart
- network plugin bringup (as briefly as possible)

Author: I think some of those may be a good follow ups

---
content_type: "reference"

> title: What Happens on Restart

Reviewer: ```suggestion
title: What Happens After A Node Restart
```

No `?`…
Personally, because this is a reference page, I think it's good to make it a little more formal.

weight: 90
---

System components on a node sometimes restart, either because of an upgrade, a
crash, or an explicit operator action. This page describes what happens to Pods

> and to the node when the [kubelet](/docs/reference/command-line-tools-reference/kubelet/),

Reviewer: nit: use a glossary tooltip for kubelet as well.

the {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}},
or the node as a whole restarts.

In a healthy cluster these restarts are usually safe and do not break running
workloads. The sections below describe the effects to be aware of, which become
more pronounced on large or heavily loaded nodes.

## Impact of a kubelet restart

When the kubelet restarts, the containers that are already running continue to
run. The kubelet re-establishes its view of the node and reconciles the running

> containers against the desired state. During this window the following happens:

Reviewer: ```suggestion
If only the kubelet restarts, the containers that are already running **continue to
run**. The kubelet re-establishes its view of the Node, and reconciles the running
containers against the desired state. During this period of time, the following happens:
```

* The kubelet re-initializes and re-synchronizes its caches, which produces a
  burst of requests to the {{< glossary_tooltip term_id="kube-apiserver" text="API server" >}}.
  On large nodes with many Pods this burst can be significant.

* The node is temporarily reported as `NotReady` until the kubelet finishes
  initializing. While the node is `NotReady`, the
  {{< glossary_tooltip term_id="kube-scheduler" text="scheduler" >}} does not
  place new Pods on it.

* The kubelet preserves the readiness of running containers across a restart.
  Each Pod's readiness drives
  {{< glossary_tooltip term_id="endpoints" text="Endpoints" >}},

>   {{< glossary_tooltip term_id="endpoint-slice" text="EndpointSlices" >}}, and

Reviewer: ```suggestion
  {{< glossary_tooltip term_id="endpoint-slice" text="EndpointSlices" >}},
  {{< glossary_tooltip term_id="endpoints" text="Endpoints" >}} and
```
The Endpoints API is deprecated; list it second.

>   Gateway configuration, so resetting it on every restart would place a large

Reviewer: ```suggestion
  downstream configuration (such as Gateways or Ingresses); this means that resetting
  container readiness on every restart would place a large
```

  load on the API server and on components that watch endpoint state, and could
  briefly remove healthy Pods from Service load balancing. This behavior is
  described in

>   [KEP-4781: Fix inconsistent container ready state after kubelet restart](https://github.com/kubernetes/enhancements/issues/4781).

Reviewer: ```suggestion
  [KEP-4781: Fix inconsistent container ready state after kubelet restart](https://www.kubernetes.dev/resources/keps/4781/)
```
(NB: the KEP metadata are wrong, but that needs fixing separately)

  Resetting container readiness to `false` on every restart was the default
  behavior for a long time. The `ChangeContainerStatusOnKubeletRestart`
  [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
  lets you revert to that behavior, but it is a deprecated legacy escape hatch
  that is slated for removal, so you should not rely on it. For more detail, see
  [Pod behavior during kubelet restarts](/docs/concepts/workloads/pods/pod-lifecycle/#kubelet-restarts).

* {{< glossary_tooltip term_id="garbage-collection" text="Garbage collection" >}}
  of unused images and containers, and Pod
  [evictions](/docs/concepts/scheduling-eviction/node-pressure-eviction/) driven
  by node-pressure, are paused for the duration of the restart and for a short
  grace period afterwards. This can slow the node's reaction to memory or disk

>   pressure.

Reviewer: ```suggestion
* During the initial kubelet startup, 
  {{< glossary_tooltip term_id="garbage-collection" text="Garbage collection" >}}
  of unused images and containers, and Pod
  [evictions](/docs/concepts/scheduling-eviction/node-pressure-eviction/) driven
  by node-pressure, are paused. This pause continues for a  a short
  grace period after the kubelet has completed its main startup routines.
  This delay can slow the node's reaction to memory or disk pressure.
```

* Ongoing image pulls are cancelled. Depending on the container runtime, a
  cancelled pull may have to start over from the beginning when it is retried.

* Pod admission runs again for the Pods on the node as the kubelet replays them

>   through its admission checks. If the node's labels or taints have changed while

Reviewer: nit: make "labels" and "taints" be glossary tooltips.

Author: done

  the kubelet was down, a Pod can fail admission and be rejected even though it
  was already running. For more detail see

>   [kubernetes/kubernetes#123859](https://github.com/kubernetes/kubernetes/issues/123859).

Reviewer: Is this an accepted bug? If so, mention in the text of this page that there is a bug (as well as hyperlinking to it for details). If there is debate about whether the behavior is buggy, we need to be cautious about how we explain it.

Overall, in a healthy cluster a kubelet restart does not break running
workloads. On large clusters with overcommitted nodes, however, the
re-initialization load and the paused garbage collection and eviction can
contribute to system instability.

## Impact of a container runtime restart

> 

Reviewer: ```suggestion

Kubernetes does not define the behavior of your
container runtime if you restart it. Depending on the container runtime
you use, a restart may trigger a stop or restart for
all local containers.
However, most container runtimes used with Kubernetes
use a configuration that allows you to restart the
runtime and leave containers executing.

```
?

Reviewer: Please respond, even if you disagree.

Author: I applied this

When the container runtime (such as
{{< glossary_tooltip term_id="containerd" text="containerd" >}} or CRI-O)
restarts, the kubelet loses its connection to the runtime until it comes back.
During this window:

* `exec` [probes](/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)
  fail for the duration of the restart, because the kubelet cannot run commands
  inside containers. With a short timeout and failure threshold, a failing
  liveness probe can cause a container to be restarted, and a failing readiness
  probe can cause the Pod to flap out of the `Ready` state.

* The node is reported as `NotReady` by the kubelet, which blocks scheduling of
  new Pods onto the node.

* Container operations such as restarts, initialization, and status updates are
  delayed until the runtime is available again.

* If an

>   [init container](/docs/concepts/workloads/pods/init-containers/) was executing

Reviewer: nit: use a glossary tooltip for "init container".

  when the runtime restarted, its execution state can be lost, in which case the
  init container runs again.

* In rare cases, interrupting an operation at a precise moment can leave state
  inconsistent:

  * An interrupted image pull may leave inconsistent image layers, which can
    render the image unusable until it is pulled again.

  * An interrupted sandbox creation, if it is terminated in the middle of a CNI
    or NRI call, may leave the sandbox in an inconsistent state, with CNI only
    partially initialized and the possibility of a resource leak.

To cause problems, a container runtime restart has to interrupt a critical
operation at a precise moment, which is a low-probability situation, so it is

> generally a safe operation. On a heavily loaded node, where every operation is

Reviewer: ```suggestion
a low risk action. On a heavily loaded node, where every operation is
```
?

slower, the window for interrupting a critical operation is larger and the
probability of hitting one of these edge cases increases.

> ## Impact of a node reboot

Reviewer: We should hyperlink to this from early in the page, and then recap how a node reboot encompasses both a container runtime restart and a kubelet restart but with more consequences.

It may seem obvious but let's be super clear.

A node reboot is the most disruptive of these events, because every container on
the node stops and the kubelet and container runtime start from scratch. When

> the node comes back:

Reviewer: ```suggestion
A node reboot is the most disruptive of these events, because every container on
the node stops. After the node boots, the kubelet and container runtime start again
with no containers actually running.
However, the kubelet retains a local memory of which Pods (and containers) were running
at the time the node was rebooted.
```

> * All containers are stopped, and the kubelet recreates them when the node comes

Reviewer: What do you mean by "All containers are stopped"? Watch out, because this clause doesn't pass the [_by zombies_](https://geoffpalmer.co.nz/writing-tip-use-zombies-to-kill-the-passive-voice/) test.

Author: rephrased, thanks

  back. Pods that stay assigned to the node, because they are not evicted or
  deleted during the reboot, are restarted in place by the kubelet, including

>   standalone Pods that are not backed by a controller. If a Pod is instead

Reviewer: ```suggestion
  _standalone Pods_ that are not backed by a
  {{< glossary_tooltip term_id="controller" text="controller" >}}.
  The kubelet continues to run existing
  {{< glossary_tooltip text="static Pods" term_id="static-pod" >}} and does
  not replace them (the kubelet may make launch containers for the existing static
  Pod).

  If an existing Pod is instead
```

  evicted or deleted, for example after the `node.kubernetes.io/not-ready`
  toleration period described below, only Pods managed by a controller such as a
  {{< glossary_tooltip term_id="deployment" text="Deployment" >}},
  {{< glossary_tooltip term_id="statefulset" text="StatefulSet" >}}, or
  {{< glossary_tooltip term_id="daemonset" text="DaemonSet" >}} get a replacement

>   Pod, on this node or elsewhere; standalone Pods are not recreated.

Reviewer: ```suggestion
  _toleration period_ described below, only Pods managed by a controller
  (such as a
  {{< glossary_tooltip term_id="deployment" text="Deployment" >}},
  {{< glossary_tooltip term_id="statefulset" text="StatefulSet" >}}, or
  {{< glossary_tooltip term_id="daemonset" text="DaemonSet" >}})
  get a replacement Pod. The replacement Pod might schedule onto a different
  node.
  
  Standalone Pods (without another object or controller managing them) are
  **not** recreated after deletion.
```

* The node registers again and is reported as `NotReady` until the kubelet,
  container runtime, and network are ready. While the node is `NotReady`, the
  node may be [tainted](/docs/concepts/scheduling-eviction/taint-and-toleration/)
  with `node.kubernetes.io/not-ready`, and after the configured toleration

>   period the control plane can evict Pods that do not tolerate it.

Reviewer: Ideally, describe the order. Does the `NotReady` get set before or after startup taints or removed? Or, perhaps, is there no interaction between startup taints and `NotReady`?

* The kubelet re-runs admission for the Pods assigned to the node, so the label
  and taint considerations described under
  [kubelet restart](#impact-of-a-kubelet-restart) apply here as well.

* For Pods that request devices, the kubelet calls the relevant
  [device plugin](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
  again to confirm the device allocations for the Pods that are being restored on
  the node. The device plugin must re-register with the kubelet after the reboot

>   so that these allocations can be reconciled.

Reviewer: What happens if the registration is delayed?

(if this isn't easy to explain, we can merge the page as-is and iterate).

Author: I rephrsed, there is not really a registration, it just a lease refresh and reconcile of status

* Local storage tied to the lifetime of a container or Pod can be lost. A
  container's writable layer is discarded when the container is recreated, so
  data written there does not survive the reboot. An
  [`emptyDir`](/docs/concepts/storage/volumes/#emptydir) volume lasts as long as
  the Pod stays on the node: a memory-backed `emptyDir` (`medium: Memory`) is
  always lost on reboot because it is held in RAM, while a disk-backed `emptyDir`
  survives a reboot as long as the Pod is not evicted or deleted, and is removed

>   only when the Pod leaves the node.

Reviewer: How does this affect the previous section about Pod lifecycle? If a Pod had an `emptyDir` memory backed volume, then the kubelet can instantly reject that Pod as soon as it can contact the API server&hellip;

&hellip;but, whatever the detail is, we should tell the reader.

Author: added more generic item above.
